docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -0,0 +1,502 @@
+would involve importing the Helm SDK (a very large dependency) into Argo CD. 
+
+### 3. Ksonnet is deprecated, and CMPs are a good place to maintain support
+
+Offloading Ksonnet to a plugin would allow us to support existing users without maintaining Ksonnet code in the more
+actively-developed base. But we need CMP parameters to provide Ksonnet support on-par with native support.
+
+### Goals
+
+Parameterized CMPs must be:
+* Easy to write
+  * An Argo CD admin should be able to write a simple parameterized CMP in just a few lines of code.
+  * An Argo CD admin should be able to write an _advanced_ parameterized CMP server relying on thorough docs.
+    
+    Writing a custom CMP server might be preferable if the parameters announcement code gets too complex to be 
+    an inline shell script.
+* Easy to install
+  * Installing a simple CMP or even a CMP with a custom server should be intuitive and painless.
+* Easy to use
+  * Argo CD end-users (for example, developers) should be able to
+    1. View and set parameters in the Argo CD Application UI
+    2. See the parameters reflected in the Application manifest
+    3. Easily read/modify the generated parameters in the manifest (they should be structured in a way that's easy to read)
+  * CMPs should be able to announce parameters with more helpful interfaces than a simple text field.
+    * For example, image parameters should be presented using the same helpful interface as the one in Kustomize applications.
+* Future-proof
+  * Since the rich parameters UI is an important feature for config management tools, the parameter definition schema 
+    should be flexible enough to announce new _types_ of parameters so the UI can customize its presentation.
+* Backwards-compatible
+  * CMPs written before this enhancement should work fine after this enhancement is released.
+  * The UI should be able to handle unknown (new) parameter types. For example, if a plugin announces a parameter of 
+    type `date`, the UI should fall back to allowing text entry. The UI can then be enhanced to provide a better input
+    mechanism in a later release.
+* Proven with a rich demonstration
+  * The initial release of this feature should include a CMP implementation of the Helm config tool. This will
+    1. Serve as a rich example for others CMP developers to mimic
+    2. Allow us to decouple the Helm config management release cycle from the Argo release cycle
+    3. Allow us to work around [this bug](https://github.com/argoproj/argo-cd/issues/7291) without including the Helm 
+       SDK in the core Argo CD code
+  * The Helm CMP must be on-par with the native implementation.
+    1. It must present an equivalent parameters UI.
+    2. It must communicate errors back to the repo-server (and then the UI) the same as the native implementation.
+
+### Non-Goals
+
+We should not:
+* Re-implement config management tools as CMPs (besides Helm)
```

Comment:
I thought we wanted implement Kustomize integration as CMP to dogfood the feature. Isn't it the case anymore?

Reply:
I could be wrong about this one. @alexmt to we want to do Kustomize-as-CMP for the initial release?